src/main/java/com/google/sps/model/GmailResponse.java

```diff
@@ -0,0 +1,288 @@
+    return gmailClient.listUserMessages(searchQuery).stream()
+        .map(
+            (message) -> {
+              try {
+                return gmailClient.getUserMessage(message.getId(), messageFormat);
+              } catch (IOException e) {
+                throw new RuntimeException(e);
+              }
+            })
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Get the amount of unread emails from the last m hours given a list of user's unread emails from
+   * last nDays. (nDays >= mHours * 24)
+   *
+   * @param unreadEmails a list of Message objects from user's gmail account. Messages must be
+   *     either FULL, METADATA, or MINIMAL MessageFormat
+   * @return number of unread emails from last mHours hours
+   * @throws GmailMessageFormatException if the Message objects do not contain an internalDate (and
+   *     are thus the wrong format)
+   */
+  private int findUnreadEmailCountHours(List<Message> unreadEmails)
+      throws GmailMessageFormatException {
+    // This is the oldest an email can be (in milliseconds since epoch) to be within the last mHours
+    long timeCutoffEpochMs = Instant.now().toEpochMilli() - mHours * 60 * 60 * 1000;
+    return (int)
+        unreadEmails.stream()
+            .filter(
+                (message) -> {
+                  if (message.getInternalDate() != 0) {
+                    return message.getInternalDate() >= timeCutoffEpochMs;
+                  }
+                  throw new GmailMessageFormatException(
+                      "Messages must be of format MINIMAL, METADATA, or FULL");
+                })
+            .count();
+  }
+
+  /**
+   * Get the amount of unread, important emails from the last nDays given a list of user's unread
+   * emails. Emails without labelIds are skipped (they may be of the correct format, but there are
+   * no label ids, which results in a null return from google for labelIds)
+   *
+   * @param unreadEmails a list of Message objects from user's gmail account. Messages must be
+   *     either FULL, METADATA, or MINIMAL MessageFormat
+   * @return number of unread, important emails from last nDays days
+   */
+  private int findImportantUnreadEmailCount(List<Message> unreadEmails) {
+    return (int)
+        unreadEmails.stream()
+            .filter(
+                (message) ->
+                    message.getLabelIds() != null && message.getLabelIds().contains("IMPORTANT"))
+            .count();
+  }
+
+  /**
+   * Get the sender of the most unread emails from the last nDays days in a user's Gmail account
+   *
+   * @param unreadEmails a list of unread emails (with METADATA OR FULL MessageFormat)
+   * @return sender of the most unread emails from the last nDays days
+   * @throws GmailMessageFormatException if the messages do not contain an internal date and/or
+   *     headers (and are thus the wrong format)
+   */
+  private String findMostFrequentSender(List<Message> unreadEmails)
```

Comment:
API calls are potentially expensive, but just looping through the data is very cheap even for large sets, so don't be afraid of looping through it a couple times if that makes things simpler. I think you could have one method that would just get the most frequent sender and another one that, given the most frequent sender, would find their most recent email.

Reply:
I actually refactored this entire function, using helper methods to make it more readable and split up the work. It's still an O(n) solution, but I think it's more readable. Let me know your thoughts!